src/Mono.Android/Android.Runtime/AndroidRuntime.cs

```diff
@@ -388,52 +388,92 @@ public override void RegisterNativeMembers (JniType nativeClass, Type type, stri
 					return;
 				}
 
-				string[] members = methods!.Split ('\n');
-				if (members.Length < 2) {
+				int newLineIndex = methods!.IndexOf ('\n');
+				if (newLineIndex == -1) {
 					if (jniAddNativeMethodRegistrationAttributePresent)
 						base.RegisterNativeMembers (nativeClass, type, methods);
 					return;
 				}
 
-				JniNativeMethodRegistration[] natives = new JniNativeMethodRegistration [members.Length-1];
-				for (int i = 0; i < members.Length; ++i) {
-					string method = members [i];
-					if (string.IsNullOrEmpty (method))
-						continue;
-					string[] toks = members [i].Split (new[]{':'}, 4);
-					Delegate callback;
-					if (toks [2] == "__export__") {
-						var mname = toks [0].Substring (2);
-						MethodInfo? minfo = null;
-						foreach (var mi in type.GetMethods (BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.Static))
-							if (mi.Name == mname && JavaNativeTypeManager.GetJniSignature (mi) == toks [1]) {
-								minfo = mi;
-								break;
+				List<JniNativeMethodRegistration> natives = new List<JniNativeMethodRegistration> (16); // an initial capacity since most classes have more than 4-8 methods
+				ReadOnlySpan<char> methodsSpan = methods;
+
+				MethodInfo []? typeMethods = null;
+				while (!methodsSpan.IsEmpty) {
+					ReadOnlySpan<char> methodLine = methodsSpan.Slice (0, newLineIndex != -1 ? newLineIndex : methodsSpan.Length);
+					if (!methodLine.IsEmpty) {
+						SplitMethodLine (methodLine,
+							out ReadOnlySpan<char> name,
+							out ReadOnlySpan<char> signature,
+							out ReadOnlySpan<char> callbackString,
+							out ReadOnlySpan<char> callbackDeclaringTypeString);
+						Delegate callback;
+						if (callbackString.SequenceEqual ("__export__")) {
+							var mname = name.Slice (2);
+							MethodInfo? minfo = null;
+							typeMethods ??= type.GetMethods (BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.Static);
+							foreach (var mi in typeMethods)
+								if (mname.SequenceEqual(mi.Name) && signature.SequenceEqual(JavaNativeTypeManager.GetJniSignature (mi))) {
```

Comment:
Please don't add `JavaNativeTypeManager.GetJniSignature(mi)` to this codepath.  It's already reflection heavy, and this would only make it *worse*:

https://github.com/xamarin/java.interop/blob/f91b0770af630cadc270290e7d6e79637ba3f11c/src/Java.Interop.Tools.TypeNameMappings/Java.Interop.Tools.TypeNameMappings/JavaNativeTypeManager.cs#L363-L372

Only the name is required.

For further rationale, build e.g. `samples/HelloWorld/HelloWorld/HelloWorld.csproj`, and then view `samples/HelloWorld//HelloWorld/obj/Debug/android/src/example/MainActivity.java`:

```java
		__md_methods = 
			"n_onCreate:(Landroid/os/Bundle;)V:GetOnCreate_Landroid_os_Bundle_Handler\n" +
			"";
		mono.android.Runtime.register ("HelloWorld.MainActivity, HelloWorld", MainActivity.class, __md_methods);
```

The method this will lookup is `GetOnCreate_Landroid_os_Bundle_Handler`.  That's *very* unique, and not likely to cause conflicts.

*Furthermore*, the types are wrong; `signature` would be `(Landroid/os/Bundle;)V`, while `GetOnCreate_Landroid_os_Bundle_Handler` (generated in e.g. `src/Mono.Android/obj/Debug/monoandroid10/android-32/mcw/Android.App.Activity.cs`) is:

```csharp
		static Delegate GetOnCreate_Landroid_os_Bundle_Handler ()
		{
			if (cb_onCreate_Landroid_os_Bundle_ == null)
				cb_onCreate_Landroid_os_Bundle_ = JNINativeWrapper.CreateDelegate ((_JniMarshal_PPL_V) n_OnCreate_Landroid_os_Bundle_);
			return cb_onCreate_Landroid_os_Bundle_;
		}
```

which would imply a "signature" of `()V` (no parameters, returns `void`), which *will rarely match*.

Reply:
Contextual related: https://github.com/xamarin/xamarin-android/wiki/Blueprint#java-type-registration